subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMutableVersionConstraint.java

```diff
@@ -131,6 +131,17 @@ public List<String> getRejectedVersions() {
        return rejectedVersions;
     }
 
+    @Override
+    public void copyFrom(VersionConstraint constraint) {
+        this.setBranch(constraint.getBranch());
+        this.require(constraint.getRequiredVersion());
+        if (!constraint.getStrictVersion().isEmpty()) { // required version must not be set to empty string
+                this.strictly(constraint.getStrictVersion());
+            }
+        this.prefer(constraint.getPreferredVersion());
+        this.reject(constraint.getRejectedVersions().toArray(new String[0]));
```

Comment:
I would be more confident reading this code if each contraint was tested for emptiness before copying.

Reply:
Agreed, also moving this code to an internal helper class to reuse it in the unrelated `DependenciesAwareVersionCatalogBuilder` that already has a similar method.